### ISSUE TITLE
False positive: FunctionalInterfaceClash after adding default method to the interface

### ISSUE THREAD

**Issue by davido** [google/error-prone#1697]
After this change: [1] we are seeing FunctionalInterfaceClash error flagged by error prone:

```
[0 / 280] [Prepa] Writing file lib/polymer_externs/polymer_closure_worker_input
[380 / 776] Action external/bazel_tools/tools/jdk/platformclasspath.jar [for host]; 3s remote-cache ... (16 actions, 15 running)
[1,354 / 2,362] Action external/bazel_tools/tools/jdk/platformclasspath.jar [for host]; 9s remote-cache ... (9 actions, 4 running)
[1,376 / 2,362] Action external/bazel_tools/tools/jdk/platformclasspath.jar [for host]; 15s remote-cache ... (29 actions running)
[1,601 / 2,362] Bundle polygerrit-ui/app/bazel-out/k8-fastbuild/bin/polygerrit-ui/app/gr-app.html.bundled.html; 16s local ... (21 actions, 20 running)
ERROR: /home/jenkins/workspace/Gerrit-verifier-bazel/gerrit/java/com/google/gerrit/lifecycle/BUILD:3:1: Building java/com/google/gerrit/lifecycle/liblifecycle.jar (2 source files) failed (Exit 1)
java/com/google/gerrit/lifecycle/LifecycleManager.java:45: error: [FunctionalInterfaceClash] When passing lambda arguments to this function, callers will need a cast to disambiguate with: add(LifecycleListener)
  public void add(RegistrationHandle handle) {
              ^
    (see https://errorprone.info/bugpattern/FunctionalInterfaceClash)
INFO: Elapsed time: 61.965s, Critical Path: 27.41s
INFO: 1032 processes: 994 remote cache hit, 38 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```
To reproduce, apply this change: [2], switch to stable-3.0 branch and run `bazel build :release`.

[1] https://gerrit-review.googlesource.com/c/gerrit/+/272812
[2] https://gerrit-review.googlesource.com/c/gerrit/+/272833

**Comment by dpursehouse** [comment]
Isn't this the expected behavior of the check?  Both of the interfaces mentioned here have a single method which has no parameters and returns void, therefore in case of a lambda the interface is ambiguous and the cast is needed.

Maybe we can "fix" this by adding `@SuppressWarnings("FunctionalInterfaceClash")` or just demoting the check to `WARN`?

I assume that if any such ambiguous situation arises, the compiler will complain.

**Comment by davido** [comment]
Yes you are right. I was confused by the fact, that the problem only showed up, after turning the second method in LifecycleListener to default:

```java
public interface LifecycleListener extends EventListener { |  
  /** Invoked when the server is starting. */
  void start();
  /** Invoked when the server is stopping. */
  // turnning this method to "default" caused error prone to flag the FunctionalInterfaceClash check
  // on LifecycleManager#add(RegistrationHandle) method, see below
  default void stop() {} 
}
```

Even though, the `RegistrationHandle` defined as:

```java
public interface RegistrationHandle {
  /** Delete this registration. */
  void remove();
}
```

However the problem is that both `add` methods in `LifecycleManager` have argument with one songle method with the same signature :

```java
/** Tracks and executes registered {@link LifecycleListener}s. */
public class LifecycleManager {
 /**
   * Add a handle that must be cleared during stop.
   *
   * @param handle the handle to add.
   */
  public void add(RegistrationHandle handle) {
    handles.add(handle);
  }

  /**
   * Add a single listener.
   *
   * @param listener the listener to add.
   */
  public void add(LifecycleListener listener) {
    listeners.add(Providers.of(listener));
  }

[...]

```

So, the problem is, that we were just lucky, that we have not had the collision from the beginning. If `LifecycleListener` would just have one single method: `start()` we would have the potential collision from the first time we migrated to Bazel or activation of  `FunctionalInterfaceClash`  bug pattern.

To disambiguate, we could rename the methods in `LifecycleManager` from just add to `addHandle`:

```java
 /**
   * Add a handle that must be cleared during stop.
   *
   * @param handle the handle to add.
   */
  public void addHandle(RegistrationHandle handle) {
    handles.add(handle);
  }

  /**
   * Add a single listener.
   *
   * @param listener the listener to add.
   */
  public void add(LifecycleListener listener) {
    listeners.add(Providers.of(listener));
  }
```

Unfortunately, caller places would have to be adjusted, also in plugins. However, we should be able do that on master anyway?

**Comment by dpursehouse** [comment]
I think we should try to engineer a situation where this clash will occur, so we can see how it can be fixed without changing the interfaces.  If it can be solved by using casts as mentioned in the warning message, I think that would be fine.  Then we can just leave this check at `WARN` severity.

**Comment by davido** [comment]
> I think we should try to engineer a situation where this clash will occur [...]

This check is only preventing hypothetical situations that may never happen, but could happen. 

However, I don't like the idee about demoting the check from error to warning. If we think that the check is not helpful, than it should be rather switched off entirly. Or, if enabled it must be left in error severity and the code must be fixed to be `FunctionalInterfaceClash` free.

As suggested, if we rename `add(RegistrationHandle handle)` to `addHandle(RegistrationHandle handle)` we are done. I coul restore my change to change abstract `stop()` method to `default` method with empty body.
